packages/Chem/src/rdkit-service/rdkit-service-worker-substructure.ts

```diff
@@ -29,95 +28,115 @@ export class RdKitServiceWorkerSubstructure extends RdKitServiceWorkerSimilarity
-            queryMol = mol;
-          } else
-            mol.delete();
-        } // else, this looks to be a real SMARTS
-      } else { // failover to queryMolBlockFailover
-        // possibly get rid of fall-over in future
-        queryMol = getMolSafe(queryMolBlockFailover, {mergeQueryHs: true}, this._rdKitModule).mol;
+  searchWithSubstructLib(queryMol: RDMol): Uint32Array {
+    const matches = new BitArray(this._substructLibrary!.size() + this._malformedIdxs!.trueCount());
+    const str = this._substructLibrary!.get_matches(queryMol, undefined, undefined, this._substructLibrary!.size());
+    const matchesIdxs = JSON.parse(str);
+    // re-calculate matches idxs considering malformed data
+    let nonMalformedCounter = 0;
+    let matchesCounter = 0;
+    for (let i = !this._malformedIdxs!.getBit(0) ? 0 : this._malformedIdxs!.findNext(0, false); i !== -1; i = this._malformedIdxs!.findNext(i, false)) {
+      if (nonMalformedCounter === matchesIdxs[matchesCounter]) {
+        matchesCounter++;
+        matches.setBit(i, true, false);
       }
+      nonMalformedCounter++;
     }
+    return matches.buffer;
+  } 
 
-    if (queryMol !== null) {
-      if (bitset) {
-        for (let i = 0; i < bitset.length; ++i) {
-          try { //get_substruct_match can potentially throw an exception, need to catch
-            // Is patternFP iff?
-            if (bitset[i] && this._rdKitMols[i] && this._rdKitMols[i]!.get_substruct_match(queryMol) !== '{}')
-              matches.push(i);
-          } catch {
-            continue;
-          }
-        }
-      } else {
-        for (let i = 0; i < this._rdKitMols!.length; ++i) {
-          try {
-            if (this._rdKitMols[i] && this._rdKitMols[i]!.get_substruct_match(queryMol) !== '{}')
-              matches.push(i);
+  searchWithPatternFps(queryMol: RDMol, molecules: string[]): Uint32Array {
+    const matches = new BitArray(molecules.length);
+    let mol: RDMol | null = null;
+    const details = JSON.stringify({sanitize: false, removeHs: false, assignStereo: false});
+    for (let i = 0; i < molecules.length; ++i) {
+          try{
+            mol = this._rdKitModule.get_mol(molecules[i], details);
+            if (mol) {
+              if (mol.get_substruct_match(queryMol) !== '{}')
+                matches.setBit(i, true);
+            }
           } catch {
             continue;
+          } finally {
+            mol?.delete();
           }
-        }
-      }
-      queryMol.delete();
-    } else
-      throw new Error('Chem | Search pattern cannot be set');
+    }
+    return matches.buffer;
+  }
 
-    return '[' + matches.join(', ') + ']';
+  searchWithoutPatternFps(queryMol: RDMol): Uint32Array {
+    const matches = new BitArray(this._rdKitMols!.length);
+    for (let i = 0; i < this._rdKitMols!.length; ++i) {
+      try {
+        if (this._rdKitMols![i] && this._rdKitMols![i]!.get_substruct_match(queryMol) !== '{}')
+          matches.setBit(i, true);
```

Comment:
No need for the notification. And actually use the setFast method everywhere in these loops, it is faster and has less parameters

Reply:
Changed to setFast where possible